File: sdks/typescript/src/usage_values.ts
/** Provider-normalized LLM usage values without owning provider execution. */
export interface TokenUsageDetails {
  readonly cachedTokens?: bigint | null;
  readonly cacheWriteTokens?: bigint | null;
  readonly reasoningTokens?: bigint | null;
}

const MAX_U32 = 0xffffffffn;
const sat = (value: bigint): bigint => value > MAX_U32 ? MAX_U32 : value;

export interface UsageInput {
  readonly promptTokens?: bigint | null;
  readonly completionTokens?: bigint | null;
  readonly totalTokens?: bigint | null;
  readonly promptTokensDetails?: TokenUsageDetails | null;
  readonly inputTokensDetails?: TokenUsageDetails | null;
  readonly outputTokensDetails?: TokenUsageDetails | null;
  readonly cacheCreationInputTokens?: bigint | null;
  readonly cacheReadInputTokens?: bigint | null;
  readonly promptCacheHitTokens?: bigint | null;
  readonly promptCacheMissTokens?: bigint | null;
}

export class Usage {
  public constructor(public readonly input: Readonly<UsageInput> = {}) {
    validateUsage(input);
  }

  public cachedPromptTokens(): bigint {
    return this.input.promptTokensDetails?.cachedTokens
      ?? this.input.inputTokensDetails?.cachedTokens
      ?? this.input.cacheReadInputTokens
      ?? this.input.promptCacheHitTokens
      ?? 0n;
  }

  public cacheCreationPromptTokens(): bigint {
    return this.input.promptTokensDetails?.cacheWriteTokens
      ?? this.input.inputTokensDetails?.cacheWriteTokens
      ?? this.input.cacheCreationInputTokens
      ?? 0n;
  }

  public effectivePromptTokens(): bigint {
    const prompt = this.input.promptTokens ?? 0n;
    if (this.input.cacheReadInputTokens != null || this.input.cacheCreationInputTokens != null) {
      return sat(prompt + this.cachedPromptTokens() + this.cacheCreationPromptTokens());
    }
    return prompt;
  }

  public effectiveTotalTokens(): bigint {
    const completion = this.input.completionTokens ?? 0n;
    if (this.input.cacheReadInputTokens != null || this.input.cacheCreationInputTokens != null) {
      return sat(this.effectivePromptTokens() + completion);
    }
    return this.input.totalTokens ?? sat(this.effectivePromptTokens() + completion);
  }

  public cacheHitRate(): number | undefined {
    const total = this.effectivePromptTokens();
    return total === 0n ? undefined : Number(this.cachedPromptTokens()) / Number(total);
  }
}

function validateUsage(input: UsageInput): void {
  const values = [
    input.promptTokens, input.completionTokens, input.totalTokens,
    input.cacheCreationInputTokens, input.cacheReadInputTokens,
    input.promptCacheHitTokens, input.promptCacheMissTokens,
    input.promptTokensDetails?.cachedTokens, input.promptTokensDetails?.cacheWriteTokens,
    input.promptTokensDetails?.reasoningTokens, input.inputTokensDetails?.cachedTokens,
    input.inputTokensDetails?.cacheWriteTokens, input.inputTokensDetails?.reasoningTokens,
    input.outputTokensDetails?.cachedTokens, input.outputTokensDetails?.cacheWriteTokens,
    input.outputTokensDetails?.reasoningTokens,
  ];
  if (values.some((value) => value != null && (typeof value !== "bigint" || value < 0n || value > MAX_U32))) {
    throw new RangeError("usage token values must fit u32");
  }
}

File: sdks/python/src/echo_agent_sdk/usage_values.py
from __future__ import annotations

from dataclasses import dataclass

MAX_U32 = (1 << 32) - 1


@dataclass(frozen=True, slots=True)
class TokenUsageDetails:
    cached_tokens: int | None = None
    cache_write_tokens: int | None = None
    reasoning_tokens: int | None = None

    def __post_init__(self) -> None:
        _validate_u32_values(
            self.cached_tokens, self.cache_write_tokens, self.reasoning_tokens
        )


@dataclass(frozen=True, slots=True)
class Usage:
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    total_tokens: int | None = None
    prompt_tokens_details: TokenUsageDetails | None = None
    input_tokens_details: TokenUsageDetails | None = None
    output_tokens_details: TokenUsageDetails | None = None
    cache_creation_input_tokens: int | None = None
    cache_read_input_tokens: int | None = None
    prompt_cache_hit_tokens: int | None = None
    prompt_cache_miss_tokens: int | None = None

    def __post_init__(self) -> None:
        _validate_u32_values(
            self.prompt_tokens,
            self.completion_tokens,
            self.total_tokens,
            self.cache_creation_input_tokens,
            self.cache_read_input_tokens,
            self.prompt_cache_hit_tokens,
            self.prompt_cache_miss_tokens,
        )

    def cached_prompt_tokens(self) -> int:
        return next(
            (
                value
                for value in (
                    self.prompt_tokens_details.cached_tokens
                    if self.prompt_tokens_details
                    else None,
                    self.input_tokens_details.cached_tokens
                    if self.input_tokens_details
                    else None,
                    self.cache_read_input_tokens,
                    self.prompt_cache_hit_tokens,
                )
                if value is not None
            ),
            0,
        )

    def cache_creation_prompt_tokens(self) -> int:
        return next(
            (
                value
                for value in (
                    self.prompt_tokens_details.cache_write_tokens
                    if self.prompt_tokens_details
                    else None,
                    self.input_tokens_details.cache_write_tokens
                    if self.input_tokens_details
                    else None,
                    self.cache_creation_input_tokens,
                )
                if value is not None
            ),
            0,
        )

    def effective_prompt_tokens(self) -> int:
        prompt = self.prompt_tokens or 0
        if (
            self.cache_read_input_tokens is not None
            or self.cache_creation_input_tokens is not None
        ):
            return min(
                MAX_U32,
                prompt
                + self.cached_prompt_tokens()
                + self.cache_creation_prompt_tokens(),
            )
        return prompt

    def effective_total_tokens(self) -> int:
        completion = self.completion_tokens or 0
        if (
            self.cache_read_input_tokens is not None
            or self.cache_creation_input_tokens is not None
        ):
            return min(MAX_U32, self.effective_prompt_tokens() + completion)
        return (
            self.total_tokens
            if self.total_tokens is not None
            else min(MAX_U32, self.effective_prompt_tokens() + completion)
        )

    def cache_hit_rate(self) -> float | None:
        total = self.effective_prompt_tokens()
        return None if total == 0 else self.cached_prompt_tokens() / total


def _validate_u32_values(*values: int | None) -> None:
    if any(
        value is not None
        and (
            isinstance(value, bool)
            or not isinstance(value, int)
            or not 0 <= value <= MAX_U32
        )
        for value in values
    ):
        raise ValueError("usage token values must fit u32")
